lib/puppet/error.rb

```diff
@@ -71,6 +71,20 @@ def to_s
       msg
     end
 
+    def to_h
+      {
+        :issue_code => issue_code,
+        :message => basic_message,
+        :full_message => to_s,
+        :file => file,
+        :line => line,
+        :pos => pos,
+        :environment => environment,
+        :node => node,
+        :arguments => arguments,
```

Comment:
I think this requires a different treatment since `arguments` can be any data type. If this error class instead complies with the protocol then you can do this:
```
      io = StringIO.new
      writer = Puppet::Pops::Serialization::JSON::Writer.new(io)
      serializer = Puppet::Pops::Serialization::Serializer.new(writer)
      serializer.write(data)
      serializer.finish
      io.string
```
to get the JSON string

Reply:
Ping @thallgren what options are needed to ensure any Ruby object gets a resonable representation in JSON when using the serializer. (As I recall it, arguments to issues can be any kind of Ruby Object).
Also wondering, if the logic in this face could use the "logging to json" support we added to support catalog preview. (Obviously it could set up logging the result to a `.json` tmp file and then render that to the console).

Reply:
I think the simplest way to resolve this would be to use the `Puppet::Pops::Serialization::ToDataConverter` to ensure that the `arguments` value is pure Data (i.e. types supported by JSON). That conversion will be a no-op for most types of arguments. The options used should probably be an empty hash (I'm making the assumption that it's not very useful to convert rich-data with "\_\_ptype__" keywords).

Reply:
I don't think that works since some issues will send a Ruby `Class` IIRC - so there will be non-data, and it needs to just use `to_s` on such things. Didn't we do this for the .json file logger in Puppet? It must do the right thing because it works... Ping @thallgren

Reply:
The `ToDataConverter` will use `to_s` on things that it doesn't recognize unless rich_data is enabled.

The .json file logger simply does this.
```
@file.puts(Puppet::Util::Json.dump(msg.to_structured_hash))
```
and to_structured_hash doesn't output arguments. It does:
```
  def to_structured_hash
    hash = {
      'level' => @level,
      'message' => @message,
      'source' => @source,
      'tags' => @tags.to_a,
      'time' => @time.iso8601(9),
    }
    %w(file line pos issue_code environment node backtrace).each do |name|
      attr_name = "@#{name}"
      hash[name] = instance_variable_get(attr_name) if instance_variable_defined?(attr_name)
    end
    hash
  end
```